### DISCUSSION TITLE
Difference between modules and buildModules

### DISCUSSION THREAD

**Discussion by phiberber** [nuxt/nuxt#18342] [Questions]
I am not sure if this is an issue, but modules and buildModules seem to be doing the same thing for me. What's the internal difference between them? Both codes below result in the same thing for me: Triggered when building but not when running the start script.

```js
modules: [
    function(options) {
      console.log('ModuleOptions', options)
    }
]
```

```js
buildModules: [
    function(options) {
      console.log('ModuleOptions', options)
    }
]
```

Edit:

Hooks don't seem to also work on production, as the following code doesn't log when executing the start script

```js
hooks: {
  "ready": () => {
    console.log('Nuxt is Ready!')
  }
},
```

**Comment by lucpotage** [comment]
The difference is explained in Nuxt 2 docs: https://nuxtjs.org/fr/docs/directory-structure/modules/

**Reply by phiberber** [reply]
It seems to have no effect in Nuxt 3 then. Unless modules only work on a dev environment.

**Reply by r9mp** [reply]
You're right. I just checked and when you do a "yarn start" in Nuxt2 the "modules" console.log is displayed (as expected). But it is not the case in Nuxt3.

**Reply by phiberber** [reply]
I'm not sure if this is actually an issue, or just something meant to be like that. Like @danielroe said, Nuxt 3 is a build time dependency. But shouldn't it follow some of the concepts of Nuxt 2?

**Comment by danielroe** [comment]
There is no difference in Nuxt 3, because Nuxt 3 is only a build time dependency.

**Reply by phiberber** [reply]
A lot of plugins like [nuxt-prune-html](https://www.npmjs.com/package/@luxdamore/nuxt-prune-html) won't work then as they need to hook the renderRoute function while in production.

**Reply by pi0** [reply]
These modules need to migrate using nitro midldeware updating response but for the time-being, on bridge, they can also leverage `generate:*` hooks.

**Reply by phiberber** [reply]
@pi0 , in the specific case of `nuxt-prune-html`. As the generate: hooks don't run on production, how would a nitro middleware know the rendered page HTML? Middlewares can't find out future responses in standard HTTP libraries. Nuxt3 uses h3, made by you; is it different in any way? 

Edit: `nuxt-prune-html` verifies if the user-agent has the text "lighthouse", if it has. It parses the SSR HTML result from `render:route`, and removes the scripts. So Lighthouse doesn't take any time running JS scripts or downloading JS bundles.

**Reply by pi0** [reply]
I guess we need to introduce some more runtime extensions to nitro then ;) (making issue soon to track)

**Reply by phiberber** [reply]
Some of the Nuxt hooks are never called. Maybe instead of introducing new runtime options to Nitro in `nuxt.config.ts`, we can just use the existent hooks? Makes it more compatible with Nuxt 2, and I don't think it's a bad design.

**Reply by pi0** [reply]
We cannot use Nuxt built time hooks in runtime. The Nitro bundle is totally decoupled. It is by design and an intentional breaking change.

**Reply by phiberber** [reply]
Don't think render:route is a built time hook, as it is never triggered.

```js
hooks: {
  "render:route": console.log
},
```

The code above prints nothing. But if it was an approved design change, it's probably recommended to move those non-working runtime hooks (all from @nuxt/server) from NuxtHooks to NitroHooks https://github.com/nuxt/framework/blob/a896c64b1232dba1d01874159a9953d5f79bccc4/packages/kit/src/types/hooks.ts#L32
https://github.com/nuxt/framework/blob/a896c64b1232dba1d01874159a9953d5f79bccc4/packages/nitro/src/context.ts#L17

**Reply by pi0** [reply]
Nitro bundle is _also_ decoupled from src/context (only src/runtime available). Indeed we need to somehow hook into nitro runtime but it is not available yet nor identical to the nuxt 2 server hook.

**Reply by phiberber** [reply]
How does Nitro receive the current hooks if it is decoupled from the Nuxt context? Wouldn't that mean that it can't access the nuxt options property?

**Reply by pi0** [reply]
It can access runtimeConfig options only. And rest should be from files in `server/` directory. But like I mentioned it is not implemented yet so I can't point to differences. Please be patient little bit more I promise you would love new extensions once added 😄
